### ISSUE TITLE
actionButton should have class argument

### ISSUE THREAD

**Issue by hadley** [rstudio/shiny#2730]
To match `downloadButton()`. (It would automatically paste `btn` to beginning of the class hierarchy)

**Comment by daattali** [comment]
I like the argument for consistency. 

I would argue that `downloadButton` should **dropp** the class argument instead of `actionButton` **adding** it. `class` is a standard attribute for these tags anyway, even without making it an explicit argument, we would still be able to call these functions in exactly the same way and I don't think it'd have any effect. None of the other possible attributes get their own parameters, I think `class` should be no different.

**Comment by nteetor** [comment]
I would second the removal of `class` from `downloadButton` as opposed to adding a formal argument to `actionButton`. `actionButton` already informally(?) supports additional classes through `...`. `downloadButton`, too, passes along `...` to the underlying builder function, the `class` argument is redundant. From the blame history, `class` was always part of `downloadButton`, `...` was added later with `class` left as a formal argument.

**Comment by hadley** [comment]
Removing `class` would mean that any attempt to set the class would override the default, which I don’t think is the correct behaviour as it would require the user to know the full class specification in order to do any customisation. 

`class` is the second most important HTML attribute (after `id` which is already an argument) and I think it would be appropriate to have it available and documented for every input and output control.

**Comment by daattali** [comment]
I'm not able to reach a computer and I very well may be wrong, but I think it might still keep the default classes. The same way that action buttons dont have a class argument right now and setting a class keeps the default ones

**Comment by nteetor** [comment]
I agree we couldn't expect a user to know or specify the default button classes. Fortunately, the htmltools builder functions concatenate duplicate attributes.

``` r
library(htmltools)

btn <- tags$button(
  class = "btn btn-default",
  class = "btn-custom"
)

print(btn)
#> <button class="btn btn-default btn-custom"></button>
```

<sup>Created on 2019-12-27 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

This feature could be used to ensure the default classes are included even if the formal `class` argument is dropped. `actionButton` currently does this. There is no `class` argument, but if one specifies a custom class the default classes are preserved and are ahead of the custom classes.

``` r
library(shiny)

ab <- actionButton(
  inputId = "count",
  label = "Label",
  class = "btn-custom"
)

print(ab)
#> <button class="btn btn-default action-button btn-custom" id="count" type="button">Label</button>
```

<sup>Created on 2019-12-27 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by wch** [comment]
@hadley, if you specify `class` for `actionButton`, it gets appended to the existing classes. For example:

```R
actionButton("asdf", "asdf",class = "foo")
#> <button class="btn btn-default action-button foo" id="asdf" type="button">asdf</button>
```

I agree with the idea that there's no need to add a `class` param to `actionButton`, since it already works. We can provide examples and better documentation showing how to add additional classes with `class`.

However, I don't think we can safely remove the `class` argument from `downloadButton`. Suppose we remove it, so the signature of `downloadButton` changes from `function (outputId, label = "Download", class = NULL, ...) ` to `function (outputId, label = "Download", ...) `. We'll need to support existing code that looks like the following:

```R
downloadButton("x", "Label", class = "foo")
downloadButton("x", "Label", "foo")
```

The first case will just work. However, the second case will no longer add `"foo"` as a class; it will append it to the text of the button, so the button will have the text "Label foo". We could choose to emit a message in that case telling users to update their code to have `class =`, but then we have the following problem: What if someone wants to use the new syntax to pass an unnamed argument as one of the `...`, like this:

```R
downloadButton("x", "Label", "extra text")
```

where they actually want the button to have the text "Label extra text"? Then we'd still be emitting the message saying to use `class = ` even though that would be wrong in this case. Anyone who _correctly_ uses the new `downloadButton` syntax with an unnamed `...` argument would get a spurious message.

To sum up, that's why I don't think we can safely remove the `class` parameter from `downloadButton`.

**Comment by daattali** [comment]
Good point re: backwards compatibility @wch

**Comment by daattali** [comment]
Perhaps in the documentation examples it would be better to set the class to something like "btn-primary" and "btn-primary btn-lg" to show actual common use cases that the user can see their effect?

**Comment by wch** [comment]
@daattali Sounds like a good idea!

**Comment by nsleep** [comment]
yep, It's a good idea. But it's still not what I expected. 
"btn-default"  prevents "btn-primary" from fully working. 
I want to use "btn-primary" to make the font appear white on a blue background, but due to the existence of "btn-default", the font appears black on a blue background.
If actionButton has no class argument, this will always be a problem, including when other R packages refer to shiny.
